proposals/3392/dataset_partitioning.md

```diff
@@ -0,0 +1,550 @@
+ * If a subset can be identified by physical locations, like directories or files (with no wildcards), use this as subset definitions. Additional logical conditions are skipped. Producer may implement a configurable maximal amount of locations collected. If this gets exceeded, this hierarchy step is skipped.
+ * We describe a subset definition by a combination of logical conditions if the approach is applicable.
+ * In case of unrecognized subset definition, a `DescriptionSubsetCondition` is used.
+
+For two first hierarchy levels, there is a possible scenario of a job reading thousands of partitions or files. 
+We allow producer to have a configurable maximal amount of partitions or locations collected, as at some point this information stops being useful. 
+
+### Examples
+
+An example subset dataset facet for SQL:
+```
+SELECT * FROM table WHERE col1 = 7 and col2 = 9
+``` 
+would be:
+```json
+{
+   "namespace": "dataset-namespace",
+   "name": "dataset-name",
+   "inputFacets": {
+     "subset": {
+       "type": "binary",
+       "left": {
+          "type": "compare", 
+          "left": {
+            "type": "field", "value": "col1"
+          },
+          "right": {
+             "type": "literal", "value": "7"
+          },
+          "comparison": "EQUAL"
+       },
+       "right": {
+          "type": "compare",
+          "left": {
+             "type": "field", "value": "col2"
+          },
+          "right": {
+             "type": "literal", "value": "9"
+          },
+          "comparison": "EQUAL"
+       },
+       "operator": "AND"
+     }
+   }
+}
+``` 
+
+An example of this would be a dataset partitioned by `business_date` and `country`. 
+A job affecting many countries for a single date would affect multiple 
+partitions and this would be described with a facet:
+
+```json
+{
+   "namespace": "dataset-namespace",
+   "name": "dataset-name",
+   "outputFacets": {
+      "subset": {
+         "type": "partition",
+         "partitions": [
+            {
+               "dimensions": {
+                  "business_date": "2024-10-15",
+                  "country": "PL"
+               }
+            },
+            {
+               "dimensions": {
+                  "business_date": "2024-10-15",
+                  "country": "DE"
+               }
+            }
+         ]
+      }
+   }
+}
+```
+
+A spark job:
+```python
+spark \
+  .read \
+  .parquet("/some/path/a=*/b=1") \
+  .filter(col("c") == 2)
+```
+
+would result in:
+
+```json
+{
+   "namespace": "dataset-namespace",
+   "name": "dataset-name",
+   "inputFacets": {
+      "subset": {
+         "type": "location",
+         "locations": [
+            "/some/path/a=1/b=1",
+            "/some/path/a=2/b=1",
+            "/some/path/a=3/b=1",
+            "/some/path/a=4/b=1"
+         ]
+      }
+   }
+}
+```
+
+If the number of locations exceeds a configurable threshold, the facet would be:
```

Comment:
It would be probably good to express that the list is not exhaustive, especially in large organizations there is a high chance that the person who sets the threshold won't be the (only) one consuming the lineage metadata, having partial data without an explicit notice might give the impression that the list is indeed complete

Reply:
Reaching the threshold should not trim the list. It will skip this subset definition type instead.

Reply:
I have misread the example, my bad. What if we added a counter of the identified locations in case they exceed the threshold? Unless you have in mind a lazy enumeration of such locations, I guess we would have this information to decide on the threshold. I think it would be interesting to know how many locations we are reading, especially when they are too many to be listed.